<h1 align="center">
  AdenineFootprinter
  <br>
</h1>

<p align="center">
  <a href="#key-features">Key Features</a> •
  <a href="#how-to-use">How To Use</a> •
  <a href="#contact">Contact</a> 

</p>


## Key Features

* Fast analysis of genome-wide nucleosome footprint using methylated adenine sites.
* Improved accuracy for nucleosome footprinting using methylated adenine sites identified by PacBio Sequencing. 
* Improved accruacy for reads with wide per-read methylation fraction.
* Default model is optimized for M.EcoGII enzyme.
* Easy to build model from user-defined control samples.

## How To Use

To clone and run this application, you'll need [pysam](https://pysam.readthedocs.io/), [Biopython](https://biopython.org/), [numpy](https://numpy.org/doc/stable/index.html), [statsmodels](https://www.statsmodels.org/dev/user-guide.html)

```bash
# Clone this repository
$ git clone 

# Go into the repository
$ cd 

# Install dependencies

# Prepare the reference

# Analyze nucleosome footprint

[Optional]
# Prepare probability model using gDNA control sample

# Analyze nucleosome footprint using user-control

[Example]
# This example analzyed nucleosome footprints in S.cerevisiae using example methylated adenine sites

# Prepare the reference

# Analyze nucleosome footprint


```

> **Note**
> pysam package is currently not availabe in windows



---
## Contact
> [David Clark Lab](https://www.nichd.nih.gov/research/atNICHD/Investigators/clark)
> 
> GitHub [@zhuweix](https://github.com/zhuweix)

